### PR TITLE
Plugin params passthru

### PR DESCRIPTION
**Pull request by rustyrussell**
This gathered some fixes along the way too...

See also: #2308

### CODE REVIEW

**Comment by conscott** [comment]
I am happy to have this PR takeover, however, one of the differences between #2308 is that I made `usage` in pylightning implicit, just based on arg names. See d14e41f9b9798b186d01ff0469e69d39b83b08b5

So if you had an rpc method declared like
```
@plugin.method("hello")
def hello(plugin, first_name, last_name, end_with="!"):
    ...
```
It would inspect the method and generate usage string
```
hello first_name last_name [ends_with]
```

Maybe if you don't explicitly pass usage, you can try to do it implicitly?

Regardless the PLUGINS.md probably needs to be updated to show how `usage` should be used in the manifest

File: contrib/plugins/helloworld.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def hello(plugin, name="world"):
     return s
 
 
-@plugin.method("init")
+@plugin.init()
```
Comment (conscott) [inline contrib/plugins/helloworld.py]: This change seems to throw an error now. 

```

**Comment by rustyrussell** [comment]
Yes, your pylightning is way way way way nicer than mine! I'll steal that...

File: contrib/pylightning/lightning/plugin.py [hunk 2]
```diff
@@ -306,9 +306,20 @@ def _getmanifest(self):
                 doc = "Undocumented RPC method from a plugin."
             doc = re.sub('\n+', ' ', doc)
 
+            argspec = inspect.getargspec(func)
+            args = argspec.args[1:]
```
Comment (conscott) [inline contrib/pylightning/lightning/plugin.py]: My assumption here was that `plugin` had to be the first arg, but @cdecker pointed out that `request` or `plugin` can be _any_ positional arg.

This need to be updated to
```
args = [arg for arg in argspec.args if arg not in ('plugin', 'request')]
```

One confusion here is that [this comment](https://github.com/ElementsProject/lightning/blob/master/contrib/pylightning/lightning/plugin.py#L64-L66)  seems to suggest `request` and `plugin` should be the last args, but that does not seem to actually matter. Maybe I am reading this wrong.

File: contrib/pylightning/lightning/plugin.py [hunk 3]
```diff
@@ -306,9 +306,20 @@ def _getmanifest(self):
...
+            defaults = argspec.defaults
+
+            # Make optional args be surrounded by square brackets
+            # list regular lightning-cli commands args
+            if defaults:
```
Comment (conscott) [inline contrib/pylightning/lightning/plugin.py]: Detecting default args below should still be correct unless someone tried to set a default for `plugin` or `request` like
```python
def hello(plugin, request=None, name="world')
```
In which case, you will know this for sure `if len(defaults) > len(args)`, in which case you could ditch method registration and print an error like "plugin and request args do not require default".

File: contrib/plugins/helloworld.py (hunk 1 above)
Comment (conscott) [inline contrib/plugins/helloworld.py]: This is actually a breaking change for existing plugin libs (i.e. mine :smile: ) 

Is that sufficient to require a version bump of pylightning ?

File: contrib/pylightning/lightning/plugin.py (hunk 2 above)
Comment (conscott) [reply contrib/pylightning/lightning/plugin.py]: Apologies for turning this commit into a :hankey: This is what you _really_ want
```python
            argspec = inspect.getargspec(func)
            defaults = argspec.defaults
            num_defaults = len(defaults) if defaults else 0
            args = []
            for idx, arg in enumerate(argspec.args):
                if arg in ('plugin', 'request'):
                    continue
                # Positional arg
                if idx < (num_defaults):
                    args.append("%s" % arg)
                # Keyword arg
                else:
                    args.append("[%s]" % arg)
```

File: contrib/pylightning/lightning/plugin.py (hunk 3 above)
Comment (conscott) [reply contrib/pylightning/lightning/plugin.py]: This is fixed by my latest comment above.

File: contrib/plugins/helloworld.py (hunk 1 above)
Comment (rustyrussell) [reply contrib/plugins/helloworld.py]: Well, plugins aren't released yet! Which is why I washed to get this right before they are...

File: contrib/pylightning/lightning/plugin.py (hunk 2 above)
Comment (rustyrussell) [reply contrib/pylightning/lightning/plugin.py]: Applied!

**Comment by rustyrussell** [comment]
OK, that change applied, and documentation is now updated too.

**Comment by conscott** [comment]
ACK 0eff8c4d1f48c3bc1baeb1ebae0e2d7a35e6bd18 :tada:

File: lightningd/memdump.c [hunk 4]
```diff
@@ -154,6 +154,7 @@ static void scan_mem(struct command *cmd,
 	memleak_remove_htable(memtable, &ld->topology->txowatches.raw);
 	memleak_remove_htable(memtable, &ld->htlcs_in.raw);
 	memleak_remove_htable(memtable, &ld->htlcs_out.raw);
+	jsonrpc_remove_memleak(memtable, ld->jsonrpc);
```
Comment (cdecker) [inline lightningd/memdump.c]: How is this not a leak?

Comment (rustyrussell) [reply lightningd/memdump.c]: These functions are kind of misnamed :(

That just removes the pointers from the table of all pointers ("memtable"), so they don't count as leaks.

Comment (cdecker) [reply lightningd/memdump.c]: But isn't `ld` supposed to be the root of allocations (and the only thing that is marked as not being a leak)? This is all reminiscent of mark and sweep garbage collection :wink:

**Pull request merged**